lib/logger/lib/logger.ex

```diff
@@ -819,11 +809,14 @@ defmodule Logger do
     end
   end
 
-  defp compile_time_purge_matching?(compile_metadata) do
+  defp compile_time_purge_matching?(level, compile_metadata) do
     matching = Application.get_env(:logger, :compile_time_purge_matching, [])
 
     Enum.any?(matching, fn filter ->
       Enum.all?(filter, fn
+        {:level_lower_than, min_level} ->
```

Comment:
I was looking at this code and I was wondering if it would be a good idea to raise an error with a specific message if the level is invalid. Right now, if the level is not `:debug`, `:info`, `:warn`, or `:error` it will raise `FunctionClauseError` in `level_to_number/1`:

https://github.com/elixir-lang/elixir/blob/90539e7c85593c00fbb4b64cfe3ada82d7a96120/lib/logger/lib/logger.ex#L527-L532

The change would be something like:

```elixir
Enum.any?(matching, fn filter ->
  Enum.all?(filter, fn
    {:level_lower_than, min_level} when min_level in @levels ->
      compare_levels(level, min_level) == :lt

    {:level_lower_than, invalid_level} ->
      raise "expected :level_lower_than in :compile_time_purge_matching to be a " <>
              "valid logging level, got: #{inspect(invalid_level)}"

    {k, v} when is_atom(k) ->
      Keyword.fetch(compile_metadata, k) == {:ok, v}

    _ ->
      raise "expected :compile_time_purge_matching to be a list of keyword lists, " <>
              "got: #{inspect(matching)}"
  end)
end)
```

Thoughts?

Reply:
Great catch!

Elixir v1.8 will require Erlang/OTP which means 99% of the FunctionClauseError will be quite readable, since we will do clause diffing. So I would like to try that first before adding extra checks.

What do you think?

Reply:
I guess you meant Erlang/OTP 20+? Ok then.

Reply:
Yes. :)